### PR TITLE
test: migrate clean to snapbox

### PR DESCRIPTION
**Pull request by dieterplex**
Part of #14039, migrating `tests/testsuite/clean.rs` to snapbox.

### CODE REVIEW

File: tests/testsuite/clean.rs [hunk 1]
```diff
@@ -38,7 +38,7 @@ fn different_dir() {
 
     p.cargo("clean")
         .cwd("src")
-        .with_stderr("[REMOVED] [..]")
+        .with_stderr_data("[REMOVED] 11 files, [FILE_SIZE]B total\n")
```
[the remark this answers is not in the document]
Comment (dieterplex) [reply tests/testsuite/clean.rs]: It does change in macos build and update to use `(PACKAGED|REMOVED|SUMMARY)` to redact in other PR.

File: tests/testsuite/clean.rs [hunk 2]
```diff
@@ -420,21 +442,25 @@ fn clean_verbose() {
...
 ",
     );
     if cfg!(target_os = "macos") {
         // Rust 1.69 has changed so that split-debuginfo=unpacked includes unpacked for rlibs.
         for obj in p.glob("target/debug/deps/bar-*.o") {
-            expected.push_str(&format!("[REMOVING] [..]{}\n", obj.unwrap().display()));
+            let obj = obj.unwrap().display().to_string();
+            expected.push_str(&format!(
+                "[REMOVING] {}\n",
+                obj.replace(p.root().to_str().unwrap(), "[ROOT]/foo")
```
Comment (dieterplex) [inline tests/testsuite/clean.rs:457]: Replacing root is not enough and there is only 1 of 2 hashes in filename being redacted. How about leave it as is for now?

```
- [REMOVING] [ROOT]/foo/target/debug/deps/bar-966252632dc060f2.bar.cce207b4909ed527-cgu.0.rcgu.o
+ [REMOVING] [ROOT]/foo/target/debug/deps/bar-[HASH].bar.cce207b4909ed527-cgu.0.rcgu.o
```

File: tests/testsuite/alt_registry.rs [hunk 3]
```diff
@@ -346,7 +346,7 @@ fn publish_with_registry_dependency() {
 See https://doc.rust-lang.org/cargo/reference/manifest.html#package-metadata for more info.
 [PACKAGING] foo v0.0.1 ([ROOT]/foo)
 [UPDATING] `alternative` index
-[PACKAGED] 4 files, [FILE_SIZE]B ([FILE_SIZE]B compressed)
```
Comment (weihanglo) [inline tests/testsuite/alt_registry.rs:349]: Was a bit unsure about redactin file count for `cargo package`. Seems they all verify the uploaded content so should be fine

File: tests/testsuite/clean.rs (hunk 2 above)
Comment (weihanglo) [reply tests/testsuite/clean.rs:457]: Yeah fine we that. I don't think Cargo cares about codegen unit artifacts. What about making it conditional, something like this?

```rust
#[cfg(not(target_os = "macos"))]
let expected = str![[r#"
[REMOVING] [ROOT]/foo/target/debug/.fingerprint/bar-[HASH]
[REMOVING] [ROOT]/foo/target/debug/deps/libbar-[HASH].rlib
[REMOVING] [ROOT]/foo/target/debug/deps/bar-[HASH].d
[REMOVING] [ROOT]/foo/target/debug/deps/libbar-[HASH].rmeta
[REMOVED] [FILE_NUM] files, [FILE_SIZE]B total

"#]];
#[cfg(target_os = "macos")]
let expected = str![[r#"
[REMOVING] [ROOT]/foo/target/debug/.fingerprint/bar-[HASH]
[REMOVING] [ROOT]/foo/target/debug/deps/libbar-[HASH].rlib
[REMOVING] [ROOT]/foo/target/debug/deps/bar-[HASH].d
[REMOVING] [ROOT]/foo/target/debug/deps/libbar-[HASH].rmeta
[REMOVING] [ROOT]/foo/target/debug/deps/bar-[HASH][..].o
[REMOVED] [FILE_NUM] files, [FILE_SIZE]B total

"#]];

Comment (dieterplex) [reply tests/testsuite/clean.rs:457]: looks promising, but is there only one .o file?

Comment (weihanglo) [reply tests/testsuite/clean.rs:457]: Nice catch. Could also just count the nubmer of `bar-.*.o` and push the same number of `[REMOVING] [ROOT]/foo/target/debug/deps/bar-[HASH][..].o` line.

File: tests/testsuite/clean.rs [hunk 4]
```diff
@@ -829,32 +847,39 @@ fn clean_dry_run() {
...
+        .with_stderr_data(str![[r#"
+[SUMMARY] [FILE_NUM] files, [FILE_SIZE]B total
+[WARNING] no files deleted due to --dry-run
+
+"#]])
         .run();
     // Verify it didn't delete anything.
     let after = p.build_dir().ls_r();
     assert_eq!(before, after);
-    let expected = itertools::join(before.iter().map(|p| p.to_str().unwrap()), "\n");
```
Comment (weihanglo) [inline tests/testsuite/clean.rs:849]: https://github.com/rust-lang/cargo/actions/runs/9590054694/job/26444813493?pr=14096#step:11:4073

Paths on Windows didn't get normalized. It also happened here <https://github.com/rust-lang/cargo/pull/14091#discussion_r1644887976>.

Seems that expected string won't be normalized? https://docs.rs/snapbox/0.6.2/src/snapbox/assert/mod.rs.html#111-141

@epage, off the top of your head, what could be the cause?


```
--- Expected
++++ actual:   stdout
   1    1 | [ROOT]/foo/target
   2      - [ROOT]/foo/target\.rustc_info.json
   3      - [ROOT]/foo/target\CACHEDIR.TAG
   4      - [ROOT]/foo/target\debug
   5      - [ROOT]/foo/target\debug\.cargo-lock
   6      - [ROOT]/foo/target\debug\.fingerprint
   7      - [ROOT]/foo/target\debug\.fingerprint\bar-72b289f5cc241a07
   8      - [ROOT]/foo/target\debug\.fingerprint\bar-72b289f5cc241a07\dep-lib-bar
   9      - [ROOT]/foo/target\debug\.fingerprint\bar-72b289f5cc241a07\invoked.timestamp
  10      - [ROOT]/foo/target\debug\.fingerprint\bar-72b289f5cc241a07\lib-bar
  11      - [ROOT]/foo/target\debug\.fingerprint\bar-72b289f5cc241a07\lib-bar.json
  12      - [ROOT]/foo/target\debug\.fingerprint\foo-ad5acdf1e6368b18
  13      - [ROOT]/foo/target\debug\.fingerprint\foo-ad5acdf1e6368b18\dep-lib-foo
  14      - [ROOT]/foo/target\debug\.fingerprint\foo-ad5acdf1e6368b18\invoked.timestamp
  15      - [ROOT]/foo/target\debug\.fingerprint\foo-ad5acdf1e6368b18\lib-foo
  16      - [ROOT]/foo/target\debug\.fingerprint\foo-ad5acdf1e6368b18\lib-foo.json
  17      - [ROOT]/foo/target\debug\build
  18      - [ROOT]/foo/target\debug\deps
  19      - [ROOT]/foo/target\debug\deps\bar-72b289f5cc241a07.d
  20      - [ROOT]/foo/target\debug\deps\foo-ad5acdf1e6368b18.d
  21      - [ROOT]/foo/target\debug\deps\libbar-72b289f5cc241a07.rmeta
  22      - [ROOT]/foo/target\debug\deps\libfoo-ad5acdf1e6368b18.rmeta
  23      - [ROOT]/foo/target\debug\examples
  24      - [ROOT]/foo/target\debug\incremental
        2 + [ROOT]/foo/target/.rustc_info.json
        3 + [ROOT]/foo/target/CACHEDIR.TAG
        4 + [ROOT]/foo/target/debug/.cargo-lock
        5 + [ROOT]/foo/target/debug/.fingerprint/bar-[HASH]/dep-lib-bar
        6 + [ROOT]/foo/target/debug/.fingerprint/bar-[HASH]/invoked.timestamp
        7 + [ROOT]/foo/target/debug/.fingerprint/bar-[HASH]/lib-bar
        8 + [ROOT]/foo/target/debug/.fingerprint/bar-[HASH]/lib-bar.json
        9 + [ROOT]/foo/target/debug/.fingerprint/bar-[HASH]
       10 + [ROOT]/foo/target/debug/.fingerprint/foo-[HASH]/dep-lib-foo
       11 + [ROOT]/foo/target/debug/.fingerprint/foo-[HASH]/invoked.timestamp
       12 + [ROOT]/foo/target/debug/.fingerprint/foo-[HASH]/lib-foo
       13 + [ROOT]/foo/target/debug/.fingerprint/foo-[HASH]/lib-foo.json
       14 + [ROOT]/foo/target/debug/.fingerprint/foo-[HASH]
       15 + [ROOT]/foo/target/debug/.fingerprint
       16 + [ROOT]/foo/target/debug/build
       17 + [ROOT]/foo/target/debug/deps/bar-[HASH].d
       18 + [ROOT]/foo/target/debug/deps/foo-[HASH].d
       19 + [ROOT]/foo/target/debug/deps/libbar-[HASH].rmeta
       20 + [ROOT]/foo/target/debug/deps/libfoo-[HASH].rmeta
       21 + [ROOT]/foo/target/debug/deps
       22 + [ROOT]/foo/target/debug/examples
       23 + [ROOT]/foo/target/debug/incremental
       24 + [ROOT]/foo/target/debug
```

Comment (dieterplex) [reply tests/testsuite/clean.rs:849]: It seems `cargo_test_support::paths::Path::ls_r()` gives non-normalized paths, but actual output.

File: crates/cargo-test-support/src/compare.rs
Comment (weihanglo) [inline crates/cargo-test-support/src/compare.rs:1]: @dieterplex could you extract this into a separate PR?
Other places like <https://github.com/rust-lang/cargo/pull/14109#discussion_r1646644214> could also benefit from this before we fix the Windows path normalization issue. Thank you.

Comment (epage) [reply crates/cargo-test-support/src/compare.rs:1]: Unresolved this since this uses #14121 and I'm concerned it might need to be reverted.

Comment (weihanglo) [reply crates/cargo-test-support/src/compare.rs:1]: See <https://github.com/rust-lang/cargo/pull/14096#discussion_r1644832897>. The PR description is updated as well <https://github.com/rust-lang/cargo/pull/14121>.

File: tests/testsuite/clean.rs (hunk 4 above)
Comment (weihanglo) [reply tests/testsuite/clean.rs:849]: @dieterplex could you revert change in this specific test case, and put a `#[allow(deprecated)]`, so that we can unblock the entire PR?

Comment (dieterplex) [reply tests/testsuite/clean.rs]: Done.

**Review by weihanglo — APPROVED** [review]

**Comment by weihanglo** [comment]
Thanks a lot!

@bors r+
